### PR TITLE
Enable PyTorch/XLA Fully Sharded Data Parallel (FSDP) for a Specific Class of Transformer Models

### PR DESCRIPTION
**Pull request by AlexWertheim**
# What does this PR do?

This PR enables the user to make use of the [PyTorch/XLA implementation of FSDP](https://github.com/pytorch/xla/tree/master/torch_xla/distributed/fsdp). Three arguments have been added to `training_args.py` to facilitate this functionality:

- `xla_fsdp`: this flag is a string containing the location of a `.json` file which specifies the FSDP arguments the user wants to use when wrapping their model.
- `xla_fsdp_nested`: this flag is a bool which determines whether each transformer block is also FSDP wrapped. Only models which expose their transformer blocks through the class attribute `transformer.h` can use this feature.
- `xla_fsdp_grad_ckpt`: this flag is a bool which determines whether gradient checkpointing is enabled for nested FSDP wrapped layers.

# Design notes and future work

1) For very large model sizes (greater than, say, 128B parameters), users may see host-side OOMs on TPUs during initialization. This can be mitigated by initializing layer weights immediately after construction, wrapping with FSDP, and moving onto the XLA device, as can be seen in [this branch](https://github.com/AlexWertheim/transformers/blob/einsum/src/transformers/models/gpt2/modeling_gpt2.py#L690-L723). We opted to enable FSDP wrapping at the trainer level, since it does not necessitate model-specific changes and does not disrupt the existing architecture for model construction and initialization. 
2) Checkpointing support for XLA FSDP is not included as part of this PR. We hope to add it soon via another PR. 
3) As indicated above, nested FSDP is only supported for models which expose their transformer blocks in a specific way. This is because naively wrapping every child layer introduces errors. There is [a PR](https://github.com/pytorch/xla/pull/4318) which will introduce auto-wrapping functionality into FSDP, and we expect that this feature will offer a much better way for all model classes to leverage nested wrapping. We also expect that auto-wrapping will enable users to perform nested wrapping multiple layers deep, which has been seen to introduce performance gains. This auto-wrap functionality needs more testing, but we hope to add this feature in a future PR.
4) We have not included testing for XLA FSDP as part of this PR. We would like to add this in a future PR. 

Thanks to @ronghanghu for his assistance in the preparation of this PR. Among other contributions, the observations that one must copy the model's forward method and replace the optimizer step are his.  

## Who can review?

@sgugger @JackCaoG

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -417,6 +417,42 @@ def __init__(
             elif FSDPOption.NO_SHARD in args.fsdp:
                 self.fsdp = ShardingStrategy.NO_SHARD
 
+        if args.xla_fsdp:
+            try:
+                from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as FSDP, checkpoint_module
+            except ImportError:
+                assert False, "Missing module XLAFullyShardedDataParallel; this module is available in torch-xla >= 1.12.0."
```
Comment (sgugger) [inline src/transformers/trainer.py:424]: Please don't use assert in code of the library, we prefer explicit exceptions.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -417,6 +417,42 @@ def __init__(
...
+            fsdp_kwargs = args.xla_fsdp_config
+            fsdp_wrap = lambda m: FSDP(m, **fsdp_kwargs)
```
Comment (sgugger) [inline src/transformers/trainer.py:426]: Let's not refactor in a lambda this for 2 uses, especially since it's so simple. Let's call FSDP with the kwargs in the 2 instances below instead.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -417,6 +417,42 @@ def __init__(
...
+            # A wrapper for gradient checkpointing
+            grad_ckpt_wrap = checkpoint_module if args.xla_fsdp_grad_ckpt else (lambda m: m)
```
Comment (sgugger) [inline src/transformers/trainer.py:428]: Same here, we prefer explicit tests even if it takes 2 more lines of code.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -417,6 +417,42 @@ def __init__(
...
+            if args.xla_fsdp_nested:
+                if not(hasattr(model, 'transformer') and hasattr(model.transformer, 'h')):
+                    raise ValueError(
+                        "Nested XLA FSDP is currently only supported for models which expose their"
+                        " transformer blocks through `transformer.h`."
+                    )
```
Comment (sgugger) [inline src/transformers/trainer.py:434]: That's just GPT/GPT-2. This is too restrictive to be merged like this in the Trainer.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR!

Adding support like this in the `Trainer` would be beneficial, but only if there is a way in FSDP XLA to do the wrapping without requiring ad-hoc code that needs to be adapted for each model. As it is structured, it will only work for GPT-2 and its copies.
